include/rcutils/testing/fault_injection.h

```diff
@@ -0,0 +1,90 @@
+// Copyright 2020 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCUTILS__TESTING_MACROS_H_
+#define RCUTILS__TESTING_MACROS_H_
+#include <stdio.h>
+
+#ifdef __cplusplus
+extern "C"
+{
+#endif
+
+void _rcutils_set_fault_injection_count(int count);
+
+int _rcutils_maybe_fail();
+
+#if defined RCUTILS_ENABLE_FAULT_INJECTION
+
+/**
+ * \def RCUTILS_MAYBE_RETURN_ERROR
+ * \brief This macro checks and decrements a static global variable atomic counter and returns
+ * `return_value_on_error` if 0.
```

Comment:
There are some examples where a failure in one location is likely to result in a failure in another location. For example, allocations may continue to fail, or file system functions may continue to fail. For the sake of coverage, there are locations where it would take two failures to reach some lines of code. Your example of cleanup is a good one.

I think there are a few of potential solutions:
1. Setting potential alternate behaviors when the counter is negative, which is what you are suggesting. I think the idea works great, it would just require writing multiple unit tests for each behavior.

2. Instead of leaving it as -1, the `RCUTILS_SET_FAULT_COUNT` macro could have a variadic version which takes as parameters counts that it's reset to after it reaches 0. 
```C
RCUTILS_SET_FAULT_COUNT(0, 0)
```
For example, would trip a second failure immediately after its first failure. This would let you test more types of failures, and then we could use this type of macro to check that code is not only one-fault tolerant, but two-fault tolerant, three-fault tolerant etc.

3. For instances where testing two particular faults is important, then we could introduce a named fault injection. Instead of `RCUTILS_CAN_RETURN_WITH_ERROR`, there would be a `RCUTILS_CAN_RETURN_WITH_ERROR_NAMED`. Then in the unit test, we could target those locations to fail explicitly in a unit test.
```C
RCUTILS_SET_FAULT_INJECTION_NAMED(...);
RCUTILS_SET_FAULT_INJECTION_NAMED(...);
... // Run test code
```
If we do introduce this one at a later date, I could see it replacing the unnamed version macro entirely. This idea requires multiple unit tests for each targeted failed injection, but is also the only option that allows for targeted injections in the first place. 

Do you have any thoughts about the other options suitability for the situation you are describing?

Reply:
I think solution (2.) would be best. We don't really know what we're hitting, so N-fault tolerant sounds as good as it gets.

Reply:
This can be added in a followup PR when the functionality becomes necessary for better coverage.